File: clients/typescript/src/conversation/stuck-detector.ts
/**
 * Stuck Detection for Conversations
 *
 * Detects when an agent is stuck in repetitive or unproductive patterns.
 * This mirrors the Python SDK's StuckDetector class.
 */

import {
  BaseEvent,
  ActionEvent,
  ObservationEvent,
  AgentErrorEvent,
  isActionEvent,
  isObservationEvent,
  isAgentErrorEvent,
  isMessageEvent,
} from '../events/types';

/**
 * Thresholds for stuck detection patterns
 */
export interface StuckDetectionThresholds {
  /** Number of identical action-observation pairs before triggering (default: 4) */
  actionObservation: number;
  /** Number of identical action-error pairs before triggering (default: 4) */
  actionError: number;
  /** Number of consecutive agent messages without user input (default: 4) */
  monologue: number;
  /** Number of alternating patterns before triggering (default: 6) */
  alternatingPattern: number;
}

/**
 * Default thresholds for stuck detection
 */
export const DEFAULT_STUCK_THRESHOLDS: StuckDetectionThresholds = {
  actionObservation: 4,
  actionError: 4,
  monologue: 4,
  alternatingPattern: 6,
};

/**
 * Result of stuck detection check
 */
export interface StuckDetectionResult {
  /** Whether the agent is stuck */
  isStuck: boolean;
  /** Type of stuck pattern detected (if any) */
  pattern?:
    | 'action_observation_loop'
    | 'action_error_loop'
    | 'monologue'
    | 'alternating_pattern'
    | 'context_window_error';
  /** Number of repetitions detected */
  repetitions?: number;
  /** Human-readable description */
  description?: string;
}

/**
 * Detects when an agent is stuck in repetitive or unproductive patterns.
 *
 * This detector analyzes the conversation history to identify various stuck patterns:
 * 1. Repeating action-observation cycles
 * 2. Repeating action-error cycles
 * 3. Agent monologue (repeated messages without user input)
 * 4. Repeating alternating action-observation patterns
 * 5. Context window errors indicating memory issues
 */
export class StuckDetector {
  private thresholds: StuckDetectionThresholds;

  constructor(thresholds?: Partial<StuckDetectionThresholds>) {
    this.thresholds = {
      ...DEFAULT_STUCK_THRESHOLDS,
      ...thresholds,
    };
  }

  /**
   * Check if the agent is stuck based on the event history.
   *
   * @param events - Array of conversation events to analyze
   * @returns StuckDetectionResult with stuck status and details
   */
  isStuck(events: BaseEvent[]): StuckDetectionResult {
    // Find the last user message index
    const lastUserMsgIndex = this.findLastUserMessageIndex(events);

    if (lastUserMsgIndex === -1) {
      // No user message found, skip detection
      return { isStuck: false };
    }

    // Only analyze events after the last user message
    const recentEvents = events.slice(lastUserMsgIndex + 1);

    // Determine minimum events needed
    const minThreshold = Math.min(
      this.thresholds.actionObservation,
      this.thresholds.actionError,
      this.thresholds.monologue
    );

    if (recentEvents.length < minThreshold) {
      return { isStuck: false };
    }

    // Collect actions and observations for analysis
    const maxNeeded = Math.max(this.thresholds.actionObservation, this.thresholds.actionError);
    const lastActions: ActionEvent[] = [];
    const lastObservations: (ObservationEvent | AgentErrorEvent)[] = [];

    // Collect from the end of history
    for (let i = recentEvents.length - 1; i >= 0; i--) {
      const event = recentEvents[i];
      if (isActionEvent(event) && lastActions.length < maxNeeded) {
        lastActions.push(event);
      } else if (
        (isObservationEvent(event) || isAgentErrorEvent(event)) &&
        lastObservations.length < maxNeeded
      ) {
        lastObservations.push(event);
      }
      if (lastActions.length >= maxNeeded && lastObservations.length >= maxNeeded) {
        break;
      }
    }

    // Check all stuck patterns

    // Scenario 1: Same action, same observation
    const actionObsResult = this.checkRepeatingActionObservation(lastActions, lastObservations);
    if (actionObsResult.isStuck) {
      return actionObsResult;
    }

    // Scenario 2: Same action, errors
    const actionErrorResult = this.checkRepeatingActionError(lastActions, lastObservations);
    if (actionErrorResult.isStuck) {
      return actionErrorResult;
    }

    // Scenario 3: Monologue
    const monologueResult = this.checkMonologue(recentEvents);
    if (monologueResult.isStuck) {
      return monologueResult;
    }

    // Scenario 4: Alternating action-observation pattern
    if (recentEvents.length >= this.thresholds.alternatingPattern) {
      const alternatingResult = this.checkAlternatingPattern(recentEvents);
      if (alternatingResult.isStuck) {
        return alternatingResult;
      }
    }

    return { isStuck: false };
  }

  /**
   * Find the index of the last user message in the events.
   */
  private findLastUserMessageIndex(events: BaseEvent[]): number {
    for (let i = events.length - 1; i >= 0; i--) {
      const event = events[i];
      if (isMessageEvent(event) && event.source === 'user') {
        return i;
      }
    }
    return -1;
  }

  /**
   * Check for repeating action-observation pattern.
   */
  private checkRepeatingActionObservation(
    lastActions: ActionEvent[],
    lastObservations: (ObservationEvent | AgentErrorEvent)[]
  ): StuckDetectionResult {
    const threshold = this.thresholds.actionObservation;

    if (lastActions.length < threshold || lastObservations.length < threshold) {
      return { isStuck: false };
    }

    // Filter to only regular observations (not errors)
    const observations = lastObservations.filter(isObservationEvent);
    if (observations.length < threshold) {
      return { isStuck: false };
    }

    // Check if all actions are identical
    const actionsEqual = lastActions
      .slice(0, threshold)
      .every((action) => this.actionsEqual(lastActions[0], action));

    // Check if all observations are identical
    const observationsEqual = observations
      .slice(0, threshold)
      .every((obs) => this.observationsEqual(observations[0], obs));

    if (actionsEqual && observationsEqual) {
      return {
        isStuck: true,
        pattern: 'action_observation_loop',
        repetitions: threshold,
        description: `Agent is repeating the same action-observation cycle ${threshold} times. Action: ${lastActions[0].tool_name}`,
      };
    }

    return { isStuck: false };
  }

  /**
   * Check for repeating action-error pattern.
   */
  private checkRepeatingActionError(
    lastActions: ActionEvent[],
    lastObservations: (ObservationEvent | AgentErrorEvent)[]
  ): StuckDetectionResult {
    const threshold = this.thresholds.actionError;

    if (lastActions.length < threshold || lastObservations.length < threshold) {
      return { isStuck: false };
    }

    // Check if all actions are identical
    const actionsEqual = lastActions
      .slice(0, threshold)
      .every((action) => this.actionsEqual(lastActions[0], action));

    if (!actionsEqual) {
      return { isStuck: false };
    }

    // Check if all observations are errors
    const allErrors = lastObservations.slice(0, threshold).every(isAgentErrorEvent);

    if (allErrors) {
      return {
        isStuck: true,
        pattern: 'action_error_loop',
        repetitions: threshold,
        description: `Agent is repeating the same action ${threshold} times and getting errors. Action: ${lastActions[0].tool_name}`,
      };
    }

    return { isStuck: false };
  }

  /**
   * Check for agent monologue (repeated messages without user input).
   */
  private checkMonologue(events: BaseEvent[]): StuckDetectionResult {
    const threshold = this.thresholds.monologue;

    if (events.length < threshold) {
      return { isStuck: false };
    }

    let agentMessageCount = 0;

    // Count consecutive agent messages from the end
    for (let i = events.length - 1; i >= 0; i--) {
      const event = events[i];

      if (isMessageEvent(event)) {
        if (event.source === 'agent') {
          agentMessageCount++;
        } else if (event.source === 'user') {
          // User interrupted, not a monologue
          break;
        }
      } else if (event.kind === 'CondensationSummaryEvent') {
        // Condensation events don't break the monologue pattern
        continue;
      } else {
        // Other events (actions/observations) break the count
        break;
      }
    }

    if (agentMessageCount >= threshold) {
      return {
        isStuck: true,
        pattern: 'monologue',
        repetitions: agentMessageCount,
        description: `Agent has sent ${agentMessageCount} consecutive messages without user input`,
      };
    }

    return { isStuck: false };
  }

  /**
   * Check for alternating action-observation pattern (A-B-A-B-A-B).
   */
  private checkAlternatingPattern(events: BaseEvent[]): StuckDetectionResult {
    const threshold = this.thresholds.alternatingPattern;

    const lastActions: ActionEvent[] = [];
    const lastObservations: (ObservationEvent | AgentErrorEvent)[] = [];

    // Collect recent actions and observations
    for (let i = events.length - 1; i >= 0; i--) {
      const event = events[i];
      if (isActionEvent(event) && lastActions.length < threshold) {
        lastActions.push(event);
      } else if (
        (isObservationEvent(event) || isAgentErrorEvent(event)) &&
        lastObservations.length < threshold
      ) {
        lastObservations.push(event);
      }
      if (lastActions.length === threshold && lastObservations.length === threshold) {
        break;
      }
    }

    if (lastActions.length < threshold || lastObservations.length < threshold) {
      return { isStuck: false };
    }

    // Check for alternating pattern: even indices match, odd indices match
    // This detects A-B-A-B-A-B patterns
    const actionsAlternate = this.checkAlternatingEquality(lastActions);
    const observationsAlternate = this.checkAlternatingEquality(lastObservations);

    if (actionsAlternate && observationsAlternate) {
      return {
        isStuck: true,
        pattern: 'alternating_pattern',
        repetitions: threshold,
        description: `Agent is stuck in an alternating action-observation pattern`,
      };
    }

    return { isStuck: false };
  }

  /**
   * Check if elements alternate (even indices match each other, odd indices match each other).
   */
  private checkAlternatingEquality<T extends BaseEvent>(items: T[]): boolean {
    if (items.length < 4) {
      return false;
    }

    // Check that items[0] == items[2] == items[4] ... and items[1] == items[3] == items[5] ...
    for (let i = 0; i < items.length - 2; i++) {
      if (!this.eventsEqual(items[i], items[i + 2])) {
        return false;
      }
    }

    return true;
  }

  /**
   * Compare two action events for equality (ignoring IDs and timestamps).
   */
  private actionsEqual(a: ActionEvent, b: ActionEvent): boolean {
    return (
      a.source === b.source &&
      a.tool_name === b.tool_name &&
      a.thought === b.thought &&
      JSON.stringify(a.action) === JSON.stringify(b.action)
    );
  }

  /**
   * Compare two observation events for equality (ignoring IDs and timestamps).
   */
  private observationsEqual(a: ObservationEvent, b: ObservationEvent): boolean {
    return (
      a.source === b.source &&
      a.tool_name === b.tool_name &&
      JSON.stringify(a.observation) === JSON.stringify(b.observation)
    );
  }

  /**
   * Compare two events for equality based on their type.
   */
  private eventsEqual(a: BaseEvent, b: BaseEvent): boolean {
    if (a.kind !== b.kind) {
      return false;
    }

    if (isActionEvent(a) && isActionEvent(b)) {
      return this.actionsEqual(a, b);
    }

    if (isObservationEvent(a) && isObservationEvent(b)) {
      return this.observationsEqual(a, b);
    }

    if (isAgentErrorEvent(a) && isAgentErrorEvent(b)) {
      return a.source === b.source && a.error === b.error;
    }

    if (isMessageEvent(a) && isMessageEvent(b)) {
      return (
        a.source === b.source && JSON.stringify(a.llm_message) === JSON.stringify(b.llm_message)
      );
    }

    // Fallback: compare JSON representation
    return JSON.stringify(a) === JSON.stringify(b);
  }
}

File: suricate-sdk/openhands/sdk/conversation/stuck_detector.py
import json

from openhands.sdk.conversation.state import ConversationState
from openhands.sdk.conversation.types import StuckDetectionThresholds
from openhands.sdk.event import (
    ActionEvent,
    AgentErrorEvent,
    CondensationSummaryEvent,
    Event,
    MessageEvent,
    ObservationBaseEvent,
    ObservationEvent,
)
from openhands.sdk.logger import get_logger


logger = get_logger(__name__)


# Maximum recent events to scan for stuck detection.
# This window should be large enough to capture repetitive patterns
# (4 repeats × 2 events per cycle = 8 events minimum, plus buffer for user messages)
# and no-progress exploration streaks (default threshold 12 actions).
MAX_EVENTS_TO_SCAN_FOR_STUCK_DETECTION: int = 40

# file_editor commands that count as making progress toward a code change
_FILE_EDITOR_PROGRESS_COMMANDS = frozenset(
    {"create", "str_replace", "insert", "undo_edit"}
)


class StuckDetector:
    """Detects when an agent is stuck in repetitive or unproductive patterns.

    This detector analyzes the conversation history to identify various stuck patterns:
    1. Repeating action-observation cycles
    2. Repeating action-error cycles
    3. Agent monologue (repeated messages without user input)
    4. Repeating alternating action-observation patterns
    5. Context window errors indicating memory issues
    """

    state: ConversationState
    thresholds: StuckDetectionThresholds

    def __init__(
        self,
        state: ConversationState,
        thresholds: StuckDetectionThresholds | None = None,
    ):
        self.state = state
        self.thresholds = thresholds or StuckDetectionThresholds()
        # Id of the AgentErrorEvent already nudged for, so a frozen streak
        # (e.g. an empty/reasoning-only response that adds no new action)
        # doesn't re-emit the same nudge every iteration.
        self._last_nudged_error_event_id: str | None = None
        # Id of the ActionEvent that completed a no-progress streak for which
        # we already emitted a nudge (one-shot until progress resets the streak).
        self._last_nudged_no_progress_action_id: str | None = None

    @property
    def action_observation_threshold(self) -> int:
        return self.thresholds.action_observation

    @property
    def action_error_threshold(self) -> int:
        return self.thresholds.action_error

    @property
    def monologue_threshold(self) -> int:
        return self.thresholds.monologue

    @property
    def alternating_pattern_threshold(self) -> int:
        return self.thresholds.alternating_pattern

    @property
    def no_progress_actions_threshold(self) -> int:
        return self.thresholds.no_progress_actions

    def _events_since_last_user_message(self) -> list[Event]:
        """Events in the scan window, after the last user message (if any).

        Windowed rather than full-history to avoid materializing large
        file-backed event logs.
        """
        events = self.state.active_branch(limit=MAX_EVENTS_TO_SCAN_FOR_STUCK_DETECTION)

        last_user_msg_index = next(
            (
                i
                for i in reversed(range(len(events)))
                if isinstance(events[i], MessageEvent) and events[i].source == "user"
            ),
            -1,  # Default to -1 if no user message found
        )
        if last_user_msg_index != -1:
            events = events[last_user_msg_index + 1 :]
        return events

    def _collect_actions_and_observations(
        self, events: list[Event], max_needed: int
    ) -> tuple[list[Event], list[Event]]:
        """The last ``max_needed`` actions and observations, most recent first."""
        last_actions: list[Event] = []
        last_observations: list[Event] = []
        for event in reversed(events):
            if isinstance(event, ActionEvent) and len(last_actions) < max_needed:
                last_actions.append(event)
            elif (
                isinstance(event, ObservationBaseEvent)
                and len(last_observations) < max_needed
            ):
                last_observations.append(event)
            if len(last_actions) >= max_needed and len(last_observations) >= max_needed:
                break
        return last_actions, last_observations

    def is_stuck(self) -> bool:
        """Check if the agent is currently stuck."""
        events = self._events_since_last_user_message()

        # Determine minimum events needed
        min_threshold = min(
            self.action_observation_threshold,
            self.action_error_threshold,
            self.monologue_threshold,
        )
        if len(events) < min_threshold:
            return False

        logger.debug(f"Checking for stuck patterns in {len(events)} events")
        logger.debug(
            f"Events after last user message: {[type(e).__name__ for e in events]}"
        )

        # action_error needs one extra pair to tell a fresh streak from one
        # that already continued past the nudge (see get_action_error_nudge)
        max_needed = max(
            self.action_observation_threshold, self.action_error_threshold + 1
        )
        last_actions, last_observations = self._collect_actions_and_observations(
            events, max_needed
        )

        # Check all stuck patterns
        # scenario 1: same action, same observation
        if self._is_stuck_repeating_action_observation(last_actions, last_observations):
            return True

        # scenario 2: same action, errors
        if self._is_stuck_repeating_action_error(last_actions, last_observations):
            return True

        # scenario 3: monologue
        if self._is_stuck_monologue(events):
            return True

        # scenario 4: action, observation alternating pattern
        if len(events) >= self.alternating_pattern_threshold:
            if self._is_stuck_alternating_action_observation(events):
                return True

        # scenario 5: context window error loop
        if len(events) >= 10:
            if self._is_stuck_context_window_error(events):
                return True

        return False

    def _is_stuck_repeating_action_observation(
        self, last_actions: list[Event], last_observations: list[Event]
    ) -> bool:
        # scenario 1: same action, same observation
        threshold = self.action_observation_threshold

        # Check for a loop of identical action-observation pairs
        if len(last_actions) >= threshold and len(last_observations) >= threshold:
            logger.debug(
                f"Found {len(last_actions)} actions and "
                f"{len(last_observations)} observations, checking for equality"
            )
            actions_equal = all(
                self._event_eq(last_actions[0], action)
                for action in last_actions[:threshold]
            )
            observations_equal = all(
                self._event_eq(last_observations[0], observation)
                for observation in last_observations[:threshold]
            )
            logger.debug(
                f"Actions equal: {actions_equal}, "
                f"Observations equal: {observations_equal}"
            )

            if actions_equal and observations_equal:
                logger.warning("Action, Observation loop detected")
                return True
        else:
            logger.debug(
                f"Not enough actions/observations: {len(last_actions)} actions,"
                f" {len(last_observations)} observations"
            )

        return False

    def _action_error_streak(
        self, last_actions: list[Event], last_observations: list[Event]
    ) -> int:
        """Length of the trailing run of one action repeatedly erroring."""
        if not last_actions or not last_observations:
            return 0
        reference = last_actions[0]
        streak = 0
        for action, observation in zip(last_actions, last_observations):
            if not self._event_eq(reference, action):
                break
            if not isinstance(observation, AgentErrorEvent):
                break
            streak += 1
        return streak

    def _is_stuck_repeating_action_error(
        self, last_actions: list[Event], last_observations: list[Event]
    ) -> bool:
        # scenario 2: same action, errors — one repeat past the threshold
        threshold = self.action_error_threshold
        if self._action_error_streak(last_actions, last_observations) > threshold:
            logger.warning("Action, Error loop detected")
            return True
        return False

    def get_action_error_nudge(self) -> str | None:
        """Nudge text once an action-error streak first hits the threshold.

        Nudges once per streak: if the streak is still frozen on the same
        error event (e.g. an empty/reasoning-only response added no new
        action) we've already nudged for it, so we don't re-fire.
        """
        events = self._events_since_last_user_message()
        threshold = self.action_error_threshold
        last_actions, last_observations = self._collect_actions_and_observations(
            events, threshold + 1
        )
        if self._action_error_streak(last_actions, last_observations) != threshold:
            return None

        action = last_actions[0]
        error = last_observations[0]
        assert isinstance(action, ActionEvent)
        assert isinstance(error, AgentErrorEvent)

        if error.id == self._last_nudged_error_event_id:
            return None
        self._last_nudged_error_event_id = error.id

        return (
            f"You've called `{action.tool_name}` with the same arguments "
            f"{threshold} times in a row and gotten the same error each "
            f"time: {error.error}. Repeating the exact same call again "
            "will not work — review the error message and either correct "
            "the arguments or try a different approach."
        )

    @staticmethod
    def _action_shows_code_progress(action: ActionEvent) -> bool:
        """True if the action is a write/finish step (not pure exploration)."""
        name = action.tool_name or ""
        if name == "finish":
            return True
        if name != "file_editor":
            return False
        raw_args = ""
        if action.tool_call is not None:
            raw_args = action.tool_call.arguments or ""
        if isinstance(raw_args, dict):
            command = raw_args.get("command")
        else:
            try:
                parsed = json.loads(raw_args) if raw_args else {}
            except (TypeError, ValueError, json.JSONDecodeError):
                return False
            command = parsed.get("command") if isinstance(parsed, dict) else None
        return command in _FILE_EDITOR_PROGRESS_COMMANDS

    def get_no_progress_nudge(self) -> str | None:
        """Nudge once after a long streak of explore-only actions.

        Unlike ``is_stuck()``, this does not stop the run — it injects a user
        message asking the agent to implement or finish. Disabled when
        ``no_progress_actions`` is 0.
        """
        threshold = self.no_progress_actions_threshold
        if threshold <= 0:
            return None

        events = self._events_since_last_user_message()
        last_actions, _ = self._collect_actions_and_observations(events, threshold)
        if len(last_actions) < threshold:
            return None

        streak = last_actions[:threshold]
        if any(self._action_shows_code_progress(a) for a in streak):
            # Progress inside the window — clear prior nudge marker so a later
            # explore streak can nudge again.
            self._last_nudged_no_progress_action_id = None
            return None

        newest = streak[0]
        assert isinstance(newest, ActionEvent)
        if newest.id == self._last_nudged_no_progress_action_id:
            return None
        self._last_nudged_no_progress_action_id = newest.id

        tool_names = [a.tool_name or "?" for a in reversed(streak)]
        return (
            f"You've taken {threshold} consecutive explore/read actions "
            f"({', '.join(tool_names[-5:])}"
            f"{', …' if len(tool_names) > 5 else ''}) without editing code or "
            "calling `finish`. Stop open-ended exploration: make a minimal "
            "code change now, or call `finish` if the task is already verified "
            "or blocked."
        )

    def _is_stuck_monologue(self, events: list[Event]) -> bool:
        # scenario 3: monologue
        # check for repeated MessageActions with source=AGENT
        # see if the agent is engaged in a good old monologue, telling
        # itself the same thing over and over
        threshold = self.monologue_threshold
        if len(events) < threshold:
            return False

        # Look for N consecutive agent messages without user interruption
        agent_message_count = 0

        for event in reversed(events):
            if isinstance(event, MessageEvent):
                if event.source == "agent":
                    agent_message_count += 1
                elif event.source == "user":
                    break  # User interrupted, not a monologue
            elif isinstance(event, CondensationSummaryEvent):
                # Condensation events don't break the monologue pattern
                continue
            else:
                # Other events (actions/observations) don't count as monologue
                break

        return agent_message_count >= threshold

    def _is_stuck_alternating_action_observation(self, events: list[Event]) -> bool:
        # scenario 4: alternating action-observation loop
        threshold = self.alternating_pattern_threshold

        last_actions: list[Event] = []
        last_observations: list[Event] = []

        # collect most recent N actions and N observations
        for event in reversed(events):
            if isinstance(event, ActionEvent) and len(last_actions) < threshold:
                last_actions.append(event)
            elif (
                isinstance(event, (ObservationEvent, AgentErrorEvent))
                and len(last_observations) < threshold
            ):
                last_observations.append(event)

            if len(last_actions) == threshold and len(last_observations) == threshold:
                break

        if len(last_actions) == threshold and len(last_observations) == threshold:
            # Check alternating pattern: [A, B, A, B, A, B] where even/odd match
            actions_equal = all(
                self._event_eq(last_actions[i], last_actions[i + 2])
                for i in range(threshold - 2)
            )
            observations_equal = all(
                self._event_eq(last_observations[i], last_observations[i + 2])
                for i in range(threshold - 2)
            )

            if actions_equal and observations_equal:
                logger.warning("Alternating Action, Observation loop detected")
                return True

        return False

    def _is_stuck_context_window_error(self, _events: list[Event]) -> bool:
        """Detects if we are stuck in a loop of context window errors.

        This happens when we repeatedly get context window errors and try to trim,
        but the trimming does not work, causing us to get more context window errors.
        The pattern is repeated AgentCondensationObservation events without any other
        events between them.
        """
        # TODO: blocked by https://github.com/yqwd-dimleap/agent-sdk/issues/282
        return False

    def _event_eq(self, event1: Event, event2: Event) -> bool:
        """
        Compare two events for equality, ignoring irrelevant
        details like ids, metrics.
        """
        # Must be same type
        if type(event1) is not type(event2):
            return False

        # For ActionEvents, compare the action content, ignoring IDs
        if isinstance(event1, ActionEvent) and isinstance(event2, ActionEvent):
            return (
                event1.source == event2.source
                and event1.thought == event2.thought
                and event1.action == event2.action
                and event1.tool_name == event2.tool_name
                # Ignore tool_call_id, llm_response_id, action_id as they vary
            )

        # For ObservationEvents, compare the observation content, ignoring IDs
        if isinstance(event1, ObservationEvent) and isinstance(
            event2, ObservationEvent
        ):
            return (
                event1.source == event2.source
                and event1.observation == event2.observation
                and event1.tool_name == event2.tool_name
                # Ignore action_id, tool_call_id as they vary
            )

        # For AgentErrorEvents, compare the error content
        if isinstance(event1, AgentErrorEvent) and isinstance(event2, AgentErrorEvent):
            return (
                event1.source == event2.source and event1.error == event2.error
                # Ignore action_id as it varies
            )

        # For MessageEvents, compare the message content
        if isinstance(event1, MessageEvent) and isinstance(event2, MessageEvent):
            return (
                event1.source == event2.source
                and event1.llm_message == event2.llm_message
            )

        # Default fallback
        return event1 == event2
